2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+        for app in self.applications:
+            for unit in app.units.with_flag(self.flags.changed):
+                host = unit.data.host
+                print('unit {} host changed: {}'.format(unit.name, host))
+                if host:
+                    unit.add_flag(self.flags.ready)
+                else:
+                    unit.remove_flag(self.flags.ready)
+                # or unit.toggle_flag(self.flags.ready, host)
+
+        for app in self.applications.without_flag(self.flags.ready):
+            print('Waiting on {}'.format(app.name))
+
+    def get_hosts(self):
+        for app in self.applications.with_flag(self.flags.ready):
+            for unit in app.units.with_flag(self.flags.ready):
+                yield unit.data.host
+```
+
+In the relatively common case that you don't expect or care about multiple
+related applications, you can use the `self.all_units` view, which is roughly
+analogous to `[u for a in self.applications for u in a.units]` except that it
+also supports the filter and flag setting / removing methods.
+
+It's important to note that each application in the collection actually
+represents a relation, and so a given remote application might actually appear
+in the collection multiple times if it is related more than once (for example,
+if an application needed multiple databases).  However, the collection of units
+and flags associated with the application or units will be kept separate.  You
+can distinguish between the application instances by their `relation_id`
+property.
+
+## Relation Data
+
+As shown above, each unit has a data attribute.  This behaves like a
+read-only `defaultdict(None)` which allows the keys to be accessed as
+attributes (with dashes converted to underscores).  The values are always
+strings or `None`.
+
+There is also a helper view for the relation data set by remote units, in case
+you expect only the leader to be providing data or for the remote units to
+agree on (some subset) of the data.  For example:
+
+```python
+class MySQLClient(SimpleRelation):
+    class flags:
+        changed = 'relation.{relation_name}.changed'
+        ready = 'relation.{relation_name}.ready'
+
+    @when(flags.changed)
+    def changed(self):
+        data = self.all_units.merged_data
+        # we assume data.host to be the master
+        if all([data.host, data.database, data.user, data.password]):
+            self.add_flag(self.flags.ready)
+```
+
+To set relation data for the current unit, there is a `set_data` collection on
+each application which behaves like the read-only data collection, but which
+allows setting attributes or dictionary values.  For example:
+
+```python
+class MySQL(SimpleRelation):
+    class flags:
+        provided = 'relation.{relation_name}.provided'
+
+    def requested_databases(self):
+        return self.applications.without_flag(self.flags.provided)
+
+    def provide_database(self, app, host, port, db, user, pass):
+        app.set_data.host = host
```

Comment:
You can have one application be related to another multiple times.  Let's say you have an application that needs two different databases for some reason.  It would have two different named relations in its `metadata.yaml` that both use the `postgres` interface protocol.  You would have to call `add-relation` twice.  On the postgresql side, it would just have the one `db` relation and that relation would have two different relation IDs that happen to point to the same remote application.

From the reactive point of view, in the postgresql interface layer, the instance would have two items in `self.applications`, one for each relation ID.  Each of those application instances would have the same application name, and the same number of units, but those unit instances would be distinct and could have different data (requesting different database names).  Data that you set on the postgresql's side of each of those relations would be independent, though all units of the remote application would see the same data (again, postgresql is saying "this is my data for this relation ID"; in reactive, the `application` instances actually map to a relation ID and maybe that means we should call them `relations` instead of `applications` but I thought that might be more confusing).